R/utils.R

```diff
@@ -497,3 +497,35 @@ generate_css_dist_html <- function(url) {
     stop(sprintf("Invalid URL supplied in external_stylesheets. Please check the syntax used for this parameter."), call. = FALSE)
 }
 
+encode_plotly <- function(layout_objs) {
+  for (i in seq_along(layout_objs)) {
+    element <- layout_objs[[i]]
+    if (is.list(element) && 
+        "x" %in% names(layout_objs) && 
+        "visdat" %in% names(layout_objs$x)) {
```

Comment:
Why is this in the loop if it's going to replace the whole `layout_objs`? Couldn't we do something like
```R
if ("x" %in% names(layout_objs) && "visdat" %in% names(layout_objs$x)) {
    obj <- suppressMessages(plotly::plotly_build(layout_objs[[something??]])$x)
    return obj[c("data", "layout")]
}
for (i in seq_along(layout_objs)) {
    element <- layout_objs[[i]]
    # the logic for the last two cases also seems backward as you have it
    # in any event you should only need to call is.list(element) once
    if (is.list(element) && any(sapply(element, is.list))) {
        layout_objs[[i]] <- encode_plotly(element)
    }
}
```

Reply:
It's a great question, and I wondered this myself, but could you elaborate about your thoughts on what `[[something??]]` might look like?

Previously, I was indexing the list using `i`, which won't exist yet in the refactored version when `plotly_build` is invoked. 

My intuition is that this change may lead to more complicated function logic. For example, if we're processing a nested list with multiple `plotly` objects, we'd need a different strategy to update (all of) them, since we no longer operate element-by-element for the `plotly_build` step.

One option could be to use the class membership, e.g. `sapply(layout_objs, function(x) "plotly" %in% class(x))`. That would return a vector of `logical()` values, which we can use to match each of the **current level's** elements that are `plotly` objects.

But that doesn't help us with the nested elements, since it looks like the revised code  has a `return` call before the for loop occurs. In that case, _I think_ we only return the top-level matches.

Am I missing something here?

Reply:
```
# the logic for the last two cases also seems backward as you have it
# in any event you should only need to call is.list(element) once
```

If it's a list that contains no other lists, and is also not a `plotly` object, we jump to the next value of `i`. If it's a nested list, we recurse. It probably makes more sense to rewrite, and omit the preceding `else if`:

```
        else if (is.list(element) & any(sapply(element, is.list))) {
            layout_objs[[i]] <- encode_plotly(element)
        }
```

I believe we still need two calls to `is.list` in the line above to check for nested list elements.

Reply:
After a very helpful discussion with @alexcjohnson, the function is cleaner and should now handle figure objects which are themselves not nested within a larger list.

Reply:
fixed in 2ca9b1587d1aea3673ef5a7376ec995e1b085b19